### PR TITLE
Minor: Add documentation + diagrams for ExternalSorter

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/5885

# Rationale for this change

While working on https://github.com/apache/arrow-datafusion/issues/5885 I have studied this code more carefully and wanted to document how it works (which is part of how I convince myself I understand the code -- if I can document it)

# What changes are included in this PR?
Doc strings and ASCII diagrams 

# Are these changes tested?
N/A

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by alamb** [comment]
@yjshen  or @Dandandan  I don't know if you might have time to review this PR, but if so I would appreciate it

File: datafusion/core/src/physical_plan/sorts/sort.rs [hunk 1]
```diff
@@ -76,27 +76,147 @@ impl ExternalSorterMetrics {
...
+///     └─────┘
+///
+///  in_mem_batches
+/// ```
 struct ExternalSorter {
+    /// schema of the output (and the input)
     schema: SchemaRef,
+    /// Potentially unsorted in memory buffer
     in_mem_batches: Vec<RecordBatch>,
+    /// if `Self::in_mem_batches` are srted
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sorts/sort.rs:201]:
```suggestion
    /// if `Self::in_mem_batches` are sorted
```

File: datafusion/core/src/physical_plan/sorts/sort.rs [hunk 2]
```diff
@@ -168,7 +288,11 @@ impl ExternalSorter {
         !self.spills.is_empty()
     }
 
-    /// MergeSort in mem batches as well as spills into total order with `SortPreservingMergeStream`.
+    /// Returns the final sorted output of all batches inserted via
+    /// [`Self::insert_batch`] as a stream of [`RecordBatch`]es.
+    ///
+    /// This may be an in memory sort/merge if all input fit into memory, or
+    /// a streaming merge from spill files on disk.
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sorts/sort.rs:295]: This process could either be an in-memory sort/merge if all the input fits into memory or a combined streaming merge incorporating both in-memory batches and data from spill files on disk.

**Review by yjshen — APPROVED** [review]
LGTM. Thanks @alamb!

File: datafusion/core/src/physical_plan/sorts/sort.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/sorts/sort.rs]: Thanks @yjshen  -- I updated the wording in 8d7d08ff0

**Pull request merged**